src/omv/imlib/draw.c

```diff
@@ -3061,19 +3020,50 @@ void imlib_draw_image(image_t *dst_img,
     size_t dst_img_row_bytes = image_size(dst_img) / dst_img->h;
 
     // Do we need to convert the image?
-    bool is_bayer_color_conversion = src_img->is_bayer && !dst_img->is_bayer;
-    bool is_yuv_color_conversion = src_img->is_yuv && !dst_img->is_yuv;
-    bool is_color_conversion = is_bayer_color_conversion || is_yuv_color_conversion;
+    bool is_bayer_conversion = src_img->is_bayer && !dst_img->is_bayer;
+    bool is_yuv_conversion = src_img->is_yuv && !dst_img->is_yuv;
+    bool is_bayer_yuv_conversion = is_bayer_conversion || is_yuv_conversion;
 
-    // Force a deep copy if we cannot use the image in-place.
-    bool need_deep_copy = (dst_img->data == src_img->data)
-                          && (is_scaling || (src_img_row_bytes < dst_img_row_bytes) || is_color_conversion);
+    // Is the line length growing which will prevent us from working in-place?
+    bool is_line_length_growing = src_img_row_bytes < dst_img_row_bytes;
 
-    // Force a deep copy if we are scaling.
-    bool is_color_conversion_scaling = is_color_conversion && is_scaling;
+    #if (OMV_GPU_ENABLE == 1)
+    if ((!callback) &&
+        (!is_bayer_yuv_conversion) &&
+        (!src_img->is_compressed) &&
+        (rgb_channel < 0) &&
+        ((dst_img->data != src_img->data) || (!is_line_length_growing)) &&
+        (!(hint & (IMAGE_HINT_AREA | IMAGE_HINT_BICUBIC | IMAGE_HINT_TRANSPOSE)))) {
```

Comment:
DMA2D completely breaks when ```dst_img->data == src_img->data && is_line_length_growing```.

For example, doing `.to_rgb565()` on an GRAYSCALE image.

`dst_img->data != src_img->data || !is_line_length_growing`  allows it to work in-place as long as the mutable format isn't growing in length or out of place. `!is_line_length_growing` by itself is unnecessarily restrictive as it prevents the GPU from being used during an out-of-place grayscale->rgb565 conversion or etc.

I could push the check into the GPU driver. But, it's somewhat universal. The GPU would need to cache the entire image or work on the pixel array backwards to avoid corrupting the image.

Reply:
True, then just rename it please.

Reply:
Done